### PR TITLE
DistCopy thrift cache

### LINKED ISSUE #38: Implement thrift cache on distcopy

### PR DESCRIPTION
**Pull request by nhibberd**
Fixes #38

### CODE REVIEW

File: notion-distcopy/src/main/java/com/ambiata/notion/distcopy/WorkloadLookup.java [hunk 1]
```diff
@@ -0,0 +1,437 @@
+/**
```
Comment (markhibberd) [inline notion-distcopy/src/main/java/com/ambiata/notion/distcopy/WorkloadLookup.java]: Going to remove this now?

File: notion-distcopy/src/main/java/com/ambiata/notion/distcopy/WorkloadsLookup.java [hunk 2]
```diff
@@ -0,0 +1,442 @@
+/**
```
Comment (markhibberd) [inline notion-distcopy/src/main/java/com/ambiata/notion/distcopy/WorkloadsLookup.java]: This one as well?

File: notion-distcopy/src/main/scala/com/ambiata/notion/distcopy/DistCopyInputFormat.scala [hunk 3]
```diff
@@ -64,16 +65,17 @@ class DistCopyInputFormat extends InputFormat[NullWritable, Mapping] {
 }
 
 object DistCopyInputFormat {
-  val MappingKey = DistCache.Key("notion.dist-copy.sync.mappings")
+  val MappingKey = ThriftCache.Key("notion.dist-copy.sync.mappings")
+
+  private val local = new java.util.concurrent.ConcurrentHashMap[String, Workloads]
+
+  def take(context: MrContext): Option[Workloads] =
```
Comment (markhibberd) [inline notion-distcopy/src/main/scala/com/ambiata/notion/distcopy/DistCopyInputFormat.scala]: This probably needs a comment, perhaps a link to the similar ivory structure, and just a justification that this is all in-memory on client side, so avoiding serialization saves memory and there is simply no way to pass it as an argument.

My bad on the Ivory one, it isn't justified well, and it looks odd, but I agree it is the best approach we have for this type of data.

**Comment by nhibberd** [comment]
@markhibberd updated

**Comment by markhibberd** [comment]
I am happy.

File: notion-distcopy/src/main/java/com/ambiata/notion/distcopy/UploadMappingLookup.java [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+        struct.setBucketIsSet(true);
+      }
+      if (incoming.get(2)) {
+        struct.key = iprot.readString();
+        struct.setKeyIsSet(true);
+      }
+    }
+  }
+
+}
```
Comment (charleso) [inline notion-distcopy/src/main/java/com/ambiata/notion/distcopy/UploadMappingLookup.java]: javaaaaaaa

File: notion-distcopy/src/main/scala/com/ambiata/notion/distcopy/DistCopyInputFormat.scala [hunk 5]
```diff
@@ -64,16 +64,22 @@ class DistCopyInputFormat extends InputFormat[NullWritable, Mapping] {
...
 object DistCopyInputFormat {
-  val MappingKey = DistCache.Key("notion.dist-copy.sync.mappings")
+  val MappingKey = ThriftCache.Key("notion.dist-copy.sync.mappings")
+
+  /*
+   This code is terrible, but a necessary evil in this instance. With the memory
+   constraints on the client, its cheaper to hold the Workloads im memory rather
+   than serializer and de-serializer them.
+   */
+  private val local = new java.util.concurrent.ConcurrentHashMap[String, Workloads]
```
Comment (charleso) [inline notion-distcopy/src/main/scala/com/ambiata/notion/distcopy/DistCopyInputFormat.scala]: o_O

**Pull request merged**